### PR TITLE
Refactor function parser

### PR DESCRIPTION
**Pull request by drwells**
Followup to #13745 - we can avoid duplicating a lot of code by adding another base class. We need to use a separate base class (instead of just functions) so that we can use the PIMPL idiom.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]
Nice simplification.

File: source/base/mu_parser_internal.cc [hunk 1]
```diff
@@ -155,53 +155,320 @@ namespace internal
...
+              "ceil",
+              "cot",
+              "csc",
+              "floor",
+              "sec",
+              "pow",
+              "erf",
+              "erfc",
+              "rand",
+              "rand_seed"};
```
Comment (masterleinad) [inline source/base/mu_parser_internal.cc:200]: Should this list be `static` and the function return a `const &` to it?

Comment (drwells) [reply source/base/mu_parser_internal.cc:200]: We could do that, but we only call this when we set up the muParser objects so its not going to make a measurable performance difference. I would prefer not to add extra static objects unless we need to.

**Comment by drwells** [comment]
*ping*

**Pull request merged**